src/System.Diagnostics.StackTrace/src/System/Diagnostics/SymbolReader.CoreCLR.cs

```diff
@@ -0,0 +1,463 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Reflection.PortableExecutable;
+using System.Runtime.InteropServices;
+
+namespace System.Diagnostics
+{
+    public class SymbolReader
+    {
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+        internal struct DebugInfo
+        {
+            public int lineNumber;
+            public int ilOffset;
+            public string fileName;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct MethodDebugInfo
+        {
+            public IntPtr points;
+            public int size;
+        }
+
+        /// <summary>
+        /// Read memory callback
+        /// </summary>
+        /// <returns>number of bytes read or 0 for error</returns>
+        internal unsafe delegate int ReadMemoryDelegate(IntPtr address, byte* buffer, int count);
+
+        /// <summary>
+        /// Internal holder for provider and reader cache.
+        /// </summary>
+        private struct HandleContents
+        {
+            readonly StackTraceSymbols _symbols;
+
+            public readonly MetadataReader PdbReader;
+
+            public HandleContents(StackTraceSymbols symbols, MetadataReader pdbReader)
+            {
+                _symbols = symbols;
+                PdbReader = pdbReader;
+            }
+
+            public void Dispose()
+            {
+                _symbols.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Stream implementation to read debugger target memory for in-memory PDBs
+        /// </summary>
+        private class TargetStream : Stream
+        {
+            readonly IntPtr _address;
+            readonly ReadMemoryDelegate _readMemory;
+
+            public override long Position { get; set; }
+            public override long Length { get; }
+            public override bool CanSeek { get { return true; } }
+            public override bool CanRead { get { return true; } }
+            public override bool CanWrite { get { return false; } }
+
+            public TargetStream(IntPtr address, int size, ReadMemoryDelegate readMemory)
+                : base()
+            {
+                _address = address;
+                _readMemory = readMemory;
+                Length = size;
+                Position = 0;
+            }
+
+            public override int Read(byte[] buffer, int offset, int count)
+            {
+                if (Position + count > Length)
+                {
+                    throw new ArgumentOutOfRangeException();
+                }
+                unsafe
+                {
+                    fixed (byte* p = &buffer[offset])
+                    {
+                        int read  = _readMemory(new IntPtr(_address.ToInt64() + Position), p, count);
```

Comment:
>_readMemory(new IntPtr(_address.ToInt64() + Position), p, count) [](start = 36, length = 64)

Is there a layering reason that prevents calling Marshal.Copy or some other reason to use the delegate?